pkg/ebpftracer/metrics.go

```diff
@@ -0,0 +1,296 @@
+package ebpftracer
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/castai/kvisor/cmd/agent/daemon/metrics"
+	"github.com/castai/kvisor/pkg/logging"
+	"github.com/cilium/ebpf"
+)
+
+const (
+	bpfStatsFile = "/proc/sys/kernel/bpf_stats_enabled"
+)
+
+var noopCleanup = func() {}
+
+// Must be kept in sync with `enum metric` defined in types.h.
+type eBPFMetric int
+
+const (
+	UnknownMetric eBPFMetric = iota
+	noFreeScratchBuffer
+	noFreeScratchBufferSocketSetState
+	noFreeScratchBufferNetflows
+	signalEventsRingbufDiscard
+	eventsRingbufDiscard
+	skbEventsRingbufDiscard
+)
+
+func (m eBPFMetric) String() string {
+	switch m {
+	case noFreeScratchBuffer:
+		return "no_free_scratch_buffer"
+	case noFreeScratchBufferSocketSetState:
+		return "no_free_scratch_buffer_socket_set_state"
+	case noFreeScratchBufferNetflows:
+		return "no_free_scratch_buffer_netflows"
+	case signalEventsRingbufDiscard:
+		return "signal_events_ringbuf_discard"
+	case eventsRingbufDiscard:
+		return "events_ringbuf_discard"
+	case skbEventsRingbufDiscard:
+		return "skb_events_ringbuf_discard"
+	default:
+		return "unknown"
+	}
+}
+
+func (t *Tracer) exportMetricsLoop(ctx context.Context) error {
+	exportTimer := time.NewTicker(t.metricExportTimerTickRate)
+	defer func() {
+		exportTimer.Stop()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-exportTimer.C:
+		}
+
+		if t.cfg.MetricsReporting.TracerMetricsEnabled {
+			if err := t.exportEBPFTracerMetrics(); err != nil {
+				t.log.Warnf("error while trying to export eBPF tracer metrics: %v", err)
+			}
+		}
+
+		if t.cfg.MetricsReporting.ProgramMetricsEnabled {
+			t.exportEBPFProgramMetrics()
+		}
+	}
+}
+
+func (t *Tracer) exportEBPFTracerMetrics() error {
+	iter := t.module.objects.tracerMaps.Metrics.Iterate()
+	metric := uint32(0)
+	counter := uint64(0)
+
+	for iter.Next(&metric, &counter) {
+		name := eBPFMetric(metric).String()
+		metrics.EBPFExposedMetrics.WithLabelValues(name).Set(float64(counter))
+		if counter > 0 {
```

Comment:
We handle the same condition inside `logInternalEbpfTracerMetric` in a way. I would get rid of this `if` as we would never log values changing to `0`.